federation/pkg/federation-controller/job/jobcontroller_test.go

```diff
@@ -0,0 +1,198 @@
+		fakeWatch.Add(obj)
+		return false, nil, nil
+	})
+	fakeClien.PrependReactor("update", resource, func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		fakeWatch.Modify(action.(core.UpdateAction).GetObject())
+		return false, nil, nil
+	})
+	fakeClien.PrependReactor("delete", resource, func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		fakeWatch.Delete(&batchv1.Job{
+			ObjectMeta: apiv1.ObjectMeta{
+				Name:      action.(core.DeleteAction).GetName(),
+				Namespace: action.GetNamespace(),
+			},
+		})
+		return false, nil, nil
+	})
+}
+
+func TestJobController(t *testing.T) {
+	flag.Set("logtostderr", "true")
+	flag.Set("v", "5")
+	flag.Parse()
+
+	jobReviewDelay = 50 * time.Millisecond
+	clusterAvailableDelay = 200 * time.Millisecond
+	clusterUnavailableDelay = 200 * time.Millisecond
+	allJobReviewDelay = 500 * time.Millisecond
+
+	fedclientset := fedclientfake.NewSimpleClientset()
+	installWatchReactor(&fedclientset.Fake, "jobs")
+
+	fedclientset.Federation().Clusters().Create(testutil.NewCluster("k8s-1", apiv1.ConditionTrue))
+	fedclientset.Federation().Clusters().Create(testutil.NewCluster("k8s-2", apiv1.ConditionTrue))
+
+	kube1clientset := kubeclientfake.NewSimpleClientset()
+	installWatchReactor(&kube1clientset.Fake, "jobs")
+	kube2clientset := kubeclientfake.NewSimpleClientset()
+	installWatchReactor(&kube2clientset.Fake, "jobs")
+
+	fedInformerClientFactory := func(cluster *fedv1.Cluster) (kubeclientset.Interface, error) {
+		switch cluster.Name {
+		case "k8s-1":
+			return kube1clientset, nil
+		case "k8s-2":
+			return kube2clientset, nil
+		default:
+			return nil, fmt.Errorf("Unknown cluster: %v", cluster.Name)
+		}
+	}
+	jobController := NewJobController(fedclientset)
+	fedjobinformer := testutil.ToFederatedInformerForTestOnly(jobController.fedJobInformer)
+	fedjobinformer.SetClientFactory(fedInformerClientFactory)
+
+	stopChan := make(chan struct{})
+	defer close(stopChan)
+	go jobController.Run(5, stopChan)
+
+	test := func(job *batchv1.Job) {
+		job, _ = fedclientset.Batch().Jobs(apiv1.NamespaceDefault).Create(job)
+		time.Sleep(1 * time.Second)
+
+		job1, _ := kube1clientset.Batch().Jobs(apiv1.NamespaceDefault).Get(job.Name)
+		finishJob(job1, 100*time.Millisecond)
+		job1, _ = kube1clientset.Batch().Jobs(apiv1.NamespaceDefault).UpdateStatus(job1)
+
+		job2, _ := kube2clientset.Batch().Jobs(apiv1.NamespaceDefault).Get(job.Name)
+		finishJob(job2, 100*time.Millisecond)
+		job2, _ = kube2clientset.Batch().Jobs(apiv1.NamespaceDefault).UpdateStatus(job2)
+
+		time.Sleep(1 * time.Second)
+		job, _ = fedclientset.Batch().Jobs(apiv1.NamespaceDefault).Get(job.Name)
+		assert.Equal(t, *job.Spec.Parallelism, *job1.Spec.Parallelism+*job2.Spec.Parallelism)
+		if job.Spec.Completions != nil {
+			assert.Equal(t, *job.Spec.Completions, *job1.Spec.Completions+*job2.Spec.Completions)
+		}
+		assert.Equal(t, job.Status.Succeeded, job1.Status.Succeeded+job2.Status.Succeeded)
```

Comment:
Also print the actual and expected succeeded if this fails:

`assert.Equal(t, job.Status.Succeeded, job1.Status.Succeeded+job2.Status.Succeeded, fmt.Sprintf("expected: %v, actual: %v", job1.Status.Succeeded+job2.Status.Succeeded, job.Status.Succeeded))`